// clean-documentation-theme/lib/components/content/section.js
'use strict';

var React = require('react');
var Radium = require('radium');

var Utils = require('../../utils');

var _require = require('../styles'),
    lineHeight = _require.lineHeight;

var SectionGroup = require('./section-group');
var SourceLink = require('./source-link');
var Extends = require('./extends');
var See = require('./see');
var Params = require('./params');
var Examples = require('./examples');
var Returns = require('./returns');
var Throws = require('./throws');

var Section = function Section(_ref) {
  var name = _ref.name,
      namespace = _ref.namespace,
      description = _ref.description,
      section = _ref.section,
      utils = _ref.utils;

  var members = section.members;
  var memberTypes = ['static', 'instance', 'events'].filter(function (key) {
    return members[key] && members[key].length > 0;
  });

  var style = {
    marginBottom: lineHeight(4)
  };

  return React.createElement(
    'div',
    { style: style },
    React.createElement(
      'h1',
      null,
      React.createElement('a', { className: 'anchor', name: utils.slug(namespace) }),
      name,
      React.createElement(SourceLink, { context: section.context })
    ),
    React.createElement('div', {
      dangerouslySetInnerHTML: {
        __html: utils.md(description)
      } }),
    React.createElement(Extends, { list: section.augments, utils: utils }),
    React.createElement(Params, { params: section.params, utils: utils }),
    React.createElement(Returns, { list: section.returns, utils: utils }),
    React.createElement(Throws, { list: section.throws, utils: utils }),
    React.createElement(Examples, { list: section.examples, utils: utils }),
    React.createElement(See, { tags: section.tags, utils: utils }),
    memberTypes.map(function (type, i) {
      return React.createElement(SectionGroup, {
        key: i,
        name: type,
        utils: utils,
        parent: name,
        members: members[type] });
    })
  );
};

Section.propTypes = {
  name: React.PropTypes.string.isRequired,
  namespace: React.PropTypes.string,
  description: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.object]).isRequired,
  section: React.PropTypes.object,
  utils: React.PropTypes.instanceOf(Utils).isRequired
};

module.exports = Radium(Section);
