// radium-bootstrap-grid/lib/column.react.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _reactPureRenderComponent = require('react-pure-render/component');

var _reactPureRenderComponent2 = _interopRequireDefault(_reactPureRenderComponent);

var _radium = require('radium');

var _radium2 = _interopRequireDefault(_radium);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _grid = require('./grid');

var Column = (function (_Component) {
  _inherits(Column, _Component);

  function Column() {
    _classCallCheck(this, _Column);

    _Component.apply(this, arguments);
  }

  Column.prototype.columnOptions = function columnOptions(size) {
    var hidden = this.props[size + 'Hidden'];
    var offset = this.props[size + 'Offset'];
    var pull = this.props[size + 'Pull'];
    var push = this.props[size + 'Push'];

    return [hidden && _grid.columnHidden[size], offset && _grid.columnOffset[size][offset], pull && _grid.columnPull[size][pull], push && _grid.columnPush[size][push]];
  };

  Column.prototype.render = function render() {
    var _props = this.props;
    var children = _props.children;
    var lg = _props.lg;
    var md = _props.md;
    var ms = _props.ms;
    var sm = _props.sm;
    var style = _props.style;
    var xs = _props.xs;

    var props = _objectWithoutProperties(_props, ['children', 'lg', 'md', 'ms', 'sm', 'style', 'xs']);

    return _react2['default'].createElement(
      'div',
      _extends({
        style: [_grid.column, _grid.columnWidth['xs'][xs || 1], _grid.columnWidth['ms'][ms || xs || 1], _grid.columnWidth['sm'][sm || ms || xs || 1], _grid.columnWidth['md'][md || sm || ms || xs || 1], _grid.columnWidth['lg'][lg || md || sm || ms || xs || 1], this.columnOptions('xs'), this.columnOptions('ms'), this.columnOptions('sm'), this.columnOptions('md'), this.columnOptions('lg'), style]
      }, props),
      children
    );
  };

  _createClass(Column, null, [{
    key: 'propTypes',
    value: {
      children: _react.PropTypes.node,
      lg: _react.PropTypes.number,
      lgHidden: _react.PropTypes.bool,
      lgOffset: _react.PropTypes.number,
      lgPull: _react.PropTypes.number,
      lgPush: _react.PropTypes.number,
      md: _react.PropTypes.number,
      mdHidden: _react.PropTypes.bool,
      mdOffset: _react.PropTypes.number,
      mdPull: _react.PropTypes.number,
      mdPush: _react.PropTypes.number,
      ms: _react.PropTypes.number,
      msHidden: _react.PropTypes.bool,
      msOffset: _react.PropTypes.number,
      msPull: _react.PropTypes.number,
      msPush: _react.PropTypes.number,
      sm: _react.PropTypes.number,
      smHidden: _react.PropTypes.bool,
      smOffset: _react.PropTypes.number,
      smPull: _react.PropTypes.number,
      smPush: _react.PropTypes.number,
      style: _react.PropTypes.object,
      xs: _react.PropTypes.number,
      xsHidden: _react.PropTypes.bool,
      xsOffset: _react.PropTypes.number,
      xsPull: _react.PropTypes.number,
      xsPush: _react.PropTypes.number
    },
    enumerable: true
  }]);

  var _Column = Column;
  Column = _radium2['default'](Column) || Column;
  return Column;
})(_reactPureRenderComponent2['default']);

exports['default'] = Column;
module.exports = exports['default'];
